### ISSUE TITLE
Mixing and matching models within a batch

### ISSUE THREAD

**Issue by zmoazeni** [bdurand/sunspot_index_queue#7]
Hey there. I believe I encountered an odd issue where processing the queue with multiple models can lose the indexable data for specific instances. This is a snippet from a local codebase:

``` ruby
> Materials::File.count
 => 1334
> Groups::Department.count
 => 1292 
> Materials::File.reindex
> Sunspot::IndexQueue.new(:session => $original_sunspot_session).process
> Materials::File.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 1334 
> Groups::Department.reindex
> Sunspot::IndexQueue.new(:session => $original_sunspot_session).process
> Groups::Department.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 1292

# All good so far. Now for the fun.

> Thread.new { Materials::File.reindex }; Thread.new { Groups::Department.reindex }
> Sunspot::IndexQueue.new(:session => $original_sunspot_session).process
> Materials::File.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 849
> Groups::Department.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 1197 

# Now process the queue 1 model at a time

> Thread.new { Materials::File.reindex }; Thread.new { Groups::Department.reindex }
> Sunspot::IndexQueue.new(:session => $original_sunspot_session, :class_names => %w(Materials::File)).process
> Sunspot::IndexQueue.new(:session => $original_sunspot_session, :class_names => %w(Groups::Department)).process
> Materials::File.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 1334 
> Groups::Department.search { paginate(:page => 1, :per_page => 10000) }.hits.count
 => 1292
```

I believe it has to do with IndexQueue#process just grabbing the next batch from Entry without segregating the entries by model.

**Comment by wakemaster39** [comment]
Wow, and here I was trying to debug this issue myself and figure out what was going on and someone solved it for me. Thanks zmoazeni, i am going to try and debug this issue some more now that I use your code to easily debug it. I will let you know if I solve it.

**Comment by wakemaster39** [comment]
applied a fix for AR in issue #8, if you use mongo or datamapper you can see how my fix works and see if you can port it. Going to try and see if there is an issue in sunspot that  could fix all of these methods at once.

**Comment by bdurand** [comment]
Fixed in pull request #9

**Comment by swalterd** [comment]
Thanks for the merge.
